trio_websocket/_impl.py

```diff
@@ -125,10 +138,33 @@ async def open_websocket(
         client-side timeout (:exc:`ConnectionTimeout`, :exc:`DisconnectionTimeout`),
         or server rejection (:exc:`ConnectionRejected`) during handshakes.
     '''
-    async with trio.open_nursery() as new_nursery:
+
+    # This context manager tries very very hard not to raise an exceptiongroup
+    # in order to be as transparent as possible for the end user.
+    # In the trivial case, this means that if user code inside the cm raises
+    # we make sure that it doesn't get wrapped.
+
+    # If opening the connection fails, then we will raise that exception. User
+    # code is never executed, so we will never have multiple exceptions.
+
+    # After opening the connection, we spawn _reader_task in the background and
+    # yield to user code. If only one of those raise a non-cancelled exception
+    # we will raise that non-cancelled exception.
+    # If we get multiple cancelled, we raise the user's cancelled.
```

Comment:
Multiple Cancelled was an existing possible case, so do we need to try hard to make it a single exception?   (Cancelled is typically caught by trio itself, and otherwise the user had to deal with MultiError Cancelled anyway.)

Reply:
The problem is that it's a very easy gotcha to write
```python
try:
  ...
except Cancelled:
  ...
```
which might work 99% of the time, but then at some random point you get multiple instead of a single (or the other way around) `Cancelled`, and your exception handler is now broken. I agree that `Cancelled` is perhaps a special case due to trio handling it, but if it does surface to the user I'm not sure there's much to be gained from having them see multiple of them.

Reply:
MultiError with Cancelled can happen fairly easily, and we quickly learned never to write `except Cancelled`.

By this trio-websocket change, we're masking the user from the effects of a trio anti-pattern, which is concerning.   Like, the user may get away with `except Cancelled` when the trio-websocket API is involved, but then be doubly puzzled when it doesn't work with another API.

Reply:
Hrm, I think this would be a very bad design choice. With this PR we're trying to make the internal nursery an implementation detail, as per https://trio.readthedocs.io/en/stable/reference-core.html#designing-for-multiple-errors
Before `strict_exception_groups=True` was the default it was expected that a trio developer would be able to handle the occasional `MultiError`, but since `strict=True` trio is trying very hard to make it fully consistent - either you always get an `ExceptionGroup`, or you never get one. And all other library APIs are encouraged to follow in those footsteps.

In the future the users shouldn't need to worry about this anti-pattern at all, so I do not think we should enshrine it in this API.